### ISSUE TITLE
AL09 Exception. TSQL aliasing expression

### ISSUE THREAD

**Issue by medianick** [sqlfluff/sqlfluff#5993]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I'm linting a T-SQL script (for SQL Server) and received the following:

```
L:  13 | P:   1 | AL09 | Unexpected exception: ;
Could you open an issue at
                       | https://github.com/sqlfluff/sqlfluff/issues ?
```

The line in question is as follows:

```sql
SELECT @CURRENT_SCHEMA_VERSION = [Version] FROM [HangFire].[Schema];
```

### Expected Behaviour

This line should not have any linting issues given the current configuration and dialect.

### Observed Behaviour

An empty exception is raised.

### How to reproduce

`sqlfluff lint Script.sql`

### Dialect

`tsql`

### Version

3.0.7

### Configuration

```ini
[sqlfluff]
dialect = tsql
exclude_rules = LT01,LT05,LT12

[sqlfluff:rules:references.quoting]
prefer_quoted_identifiers = True

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
additional_allowed_characters = .
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by medianick** [comment]
Adding the suggested `-- noqa: AL09` on the problematic line results in this:

```
CRITICAL   [AL09] Applying rule AL09 to 'Redacted/Script.sql' threw an Exception:
Traceback (most recent call last):
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/core/rules/base.py", line 507, in crawl
    res = self._eval(context=context)
  File "/app/.venv/lib/python3.9/site-packages/sqlfluff/rules/aliasing/AL09.py", line 99, in _eval
    assert whitespace and column_identifier and alias_identifier
AssertionError
```

So, this is a blocker for linting.

**Comment by medianick** [comment]
Looking over the code in https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/rules/aliasing/AL09.py, it appears that this line is being incorrectly detected as containing a _column alias_ (e.g., `select col1 as al1`) when in fact it's assigning a column into a variable. I expect this is due to SQL Server (and SQLFluff by extension) supporting the old `SELECT 'Foo' = Bar` syntax for column aliases (see https://www.red-gate.com/hub/product-learning/sql-prompt/sql-prompt-code-analysis-avoid-non-standard-column-aliases). It would be good to allow that to be opted out of or into via configuration switch, or to adjust the alias detection so that this is correctly detected as the variable assignment it is, rather than a column alias.

**Comment by alanmcruickshank** [comment]
@medianick - thanks for looking into this issue. I'm in the process of resolving some issues with AL09 anyway so this is very useful.

I'm not entirely familiar with TSQL in this regard, but how would you suggest that SQLFluff should distinguish between an old school aliasing expression and a variable assignation expression (without knowing what configuration someone has enabled in their database)?

Would you also be in favour of a TSQL specific rule here to detect an old-style aliasing expression and mark it as a linting error?

**Comment by medianick** [comment]
> how would you suggest that SQLFluff should distinguish between an old school aliasing expression and a variable assignation expression 

It should be as simple as detecting whether it's assigning to an _unquoted_ variable (starting with `@`) on the left side of the expression, e.g.:

```sql
declare @foo int

create table #Foo (Bar int)

insert into #Foo (Bar)
values (1), (2)

-- Sybase-style column alias as "@foo" via assignment; these each return 2 rows
select '@foo' = Bar from #Foo
select "@foo" = Bar from #Foo
select [@foo] = Bar from #Foo

-- Variable assignment into @foo; this doesn't return anything
select @foo = Bar from #Foo
 
drop table #Foo
```

To start an alias with `@`, it would have to be quoted with single quotes, double quotes, or square brackets. Without them, it's an assignment.

> Would you also be in favour of a TSQL specific rule here to detect an old-style aliasing expression and mark it as a linting error?

Yes, definitely; it's a legacy of Sybase that SQL Server picked up, and other tools (like Redgate's SQL Code Analysis, which I linked to above) flag it as a style violation if found.

**Comment by ariademur** [comment]
First of all thanks for the amazing tool, really great work!

I am also having the same "AL09" "Unexpected exception".

Code is a select that contains an alias, for example:

```sql
SELECT users.email as 'Email'
```

Error:

```shell
CRITICAL   [AL09] Applying rule AL09 to 'FILE.sql' threw an Exception:                                                                                                    
Traceback (most recent call last):
  File "/opt/venv/lib/python3.11/site-packages/sqlfluff/core/rules/base.py", line 507, in crawl
    res = self._eval(context=context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/sqlfluff/rules/aliasing/AL09.py", line 99, in _eval
    assert whitespace and column_identifier and alias_identifier
AssertionError
== [FILE.sql] FAIL                             
L:   4 | P:   1 | AL09 | Unexpected exception: ;                                                                           
Could you open an issue at
                       | https://github.com/sqlfluff/sqlfluff/issues ?
You can
                       | ignore this exception for now, by adding '-- noqa: AL09'
                       | at the end
of line 4

                       | [aliasing.self_alias.column]

```

Removing the alias makes the error go away.

Info:
- command: `sqlfluff lint --dialect mysql PATH`
- version: 3.1.1
- dialect: mysql
- no special configuration

**Comment by medianick** [comment]
@ariademur I think your finding should be raised as a separate issue; it's the same AL09 exception but an entirely different dialect (mysql vs. tsql) and appears to be a different syntax in the SQL script that triggers it (where my issue was triggered by the Sybase-style column aliases through '=', not 'as').

**Comment by ariademur** [comment]
Thanks, I opened a separate issue: https://github.com/sqlfluff/sqlfluff/issues/6218

**Comment by alanmcruickshank** [comment]
I'm going to close this issue as resolved. The query in the original post here no longer raises and exception or triggers AL09.

**Comment by medianick** [comment]
Confirmed here as well, with the latest 3.2.3 version. Now that it no longer throws, running this process to completion now uncovers a larger set of "Found unparsable section" linting errors -- but not exceptions, thankfully -- so I'll go through those and see if I can make sense of the cause. I'll open an issue if I can discover what triggers those, as they're valid T-SQL -- SQL Server understands it, but Sqlfluff does not.

**Comment by medianick** [comment]
https://github.com/sqlfluff/sqlfluff/issues/6339 is the first of the issues I mentioned.

**Comment by medianick** [comment]
And https://github.com/sqlfluff/sqlfluff/issues/6340 is the next.
